File: Pod/Classes/Swocket.swift
// The MIT License (MIT)
//
// Copyright (c) 2015 Joakim Gyllström
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Foundation

public enum SwocketError: ErrorType {
    case Connection
    case Send
    case Recieve
}

public typealias SwocketAsyncError = (error: SwocketError) -> ()

public class Swocket { // TODO: Struct?
    private var sockfd: Int32?
    private let port: [CChar]
    private let host: [CChar]
    private let dispatchQueue: dispatch_queue_t
    private let callbackQueue: dispatch_queue_t
    
    // MARK: Public functions
    public init(port: UInt, host: String, callback: dispatch_queue_t = dispatch_get_main_queue()) {
        // Forcefull unwrapp on purpose
        self.port = String(port).cStringUsingEncoding(NSUTF8StringEncoding)!
        self.host = host.cStringUsingEncoding(NSUTF8StringEncoding)!
        self.callbackQueue = callback
        dispatchQueue = dispatch_queue_create("\(host):\(port)", nil)
    }
    
    deinit {
        // Make sure we close connection on deallocation
        disconnect()
    }
    
    /**
    Listen for incomming connections on a given port.
    - Parameter port: The port to listen for connections on.
    - Parameter onConnection: A closure to handle incoming connections from.
    - Returns: The server socket.
    */
    public class func listen(port: UInt, onConnection: (server: Swocket, client: Swocket) -> (), error: SwocketAsyncError? = nil) -> Swocket {
        
        return Swocket(port: port, host: "")
    }
    
    /**
    Send data through the socket.
    - Parameter send: A closure that gets called when socket is ready to send. Should return the NSData to send.
    - Returns: This socket so you can chain operations
    */
    public final func send(data: NSData, error: SwocketAsyncError? = nil) -> Swocket {
        dispatch_async(dispatchQueue) { () -> Void in
            if let sockfd = self.sockfd {
                
                var totalSent = 0
                var bytesLeft = data.length
                var sentChunkSize = 0
                
                // Repeat until all bytes are sent, or we get an error
                repeat {
                    sentChunkSize = Foundation.send(sockfd, data.bytes+totalSent, bytesLeft, 0)
                    if sentChunkSize == -1 {
                        if let error = error {
                            error(error: SwocketError.Send)
                        }
                        break
                    }
                    
                    totalSent += sentChunkSize
                    bytesLeft -= sentChunkSize
                } while totalSent < data.length
            } else {
                if let error = error {
                    error(error: SwocketError.Send)
                }
            }
        }
        
        return self
    }
    
    /**
    Recieve data from a socket.
    - Parameter recieve: Closure that get called with the recieving data.
    - Returns: This socket, so you can chain operations
    */
    public final func recieve(recieve: (socket: Swocket, data: NSData) -> (), error: SwocketAsyncError? = nil) -> Swocket {
        
        dispatch_async(dispatchQueue) { () -> Void in
            if let sockfd = self.sockfd {
                let maxSize = 100
                var nullChar: Int8 = 0
                let data = NSMutableData(bytes: &nullChar, length: maxSize)
                let dataPtr = UnsafeMutablePointer<Void>(data.mutableBytes)
                
                let numBytes = recv(sockfd, dataPtr, maxSize-1, 0)
                
                dispatch_async(self.callbackQueue, { () -> Void in
                    recieve(socket: self, data: NSData(bytes: data.bytes, length: numBytes))
                })
            } else {
                // ERROR!
            }
        }
        
        return self
    }
    
    /**
    Close connection.
    */
    public final func disconnect() {
        dispatch_async(dispatchQueue) { () -> Void in
            if let sockfd = self.sockfd {
                // Close socket
                close(sockfd)
                self.sockfd = nil
            }
        }
    }
    
    // MARK: Private functions
    /**
    */
    public final func connect(error: SwocketAsyncError? = nil) {
        dispatch_async(dispatchQueue) { () -> Void in
            // Try to connect, -1 indicates failure
            let newFd = swocket_connect(self.port, self.host)
            if let error = error where newFd == -1 {
                error(error: SwocketError.Connection)
            }
            
            // Assign socket descriptor
            self.sockfd = newFd
        }
    }
}

File: Source/Swocket.c
// The MIT License (MIT)
//
// Copyright (c) 2015 Joakim Gyllström
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "Swocket.h"

void sigchld_handler(int s);
void *get_in_addr(struct sockaddr *sa);

void sigchld_handler(int s) {
    while(waitpid(-1, NULL, WNOHANG) > 0);
}

void *get_in_addr(struct sockaddr *sa) {
    if (sa->sa_family == AF_INET) {
        return &(((struct sockaddr_in*)sa)->sin_addr);
    }
    return &(((struct sockaddr_in6*)sa)->sin6_addr);
}

int swocket_open(const char * port, const char * host) {
    int sockfd;
    struct addrinfo hints, *servinfo, *p;
    int rv;
    char s[INET6_ADDRSTRLEN];
    
    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    if ((rv = getaddrinfo(host, port, &hints, &servinfo)) != 0) {
        return -1;
    }
    
    // loop through all the results and connect to the first we can
    for(p = servinfo; p != NULL; p = p->ai_next) {
        if ((sockfd = socket(p->ai_family, p->ai_socktype,
                             p->ai_protocol)) == -1) {
            continue; }
        if (connect(sockfd, p->ai_addr, p->ai_addrlen) == -1) {
            close(sockfd);
            continue; }
        break; }
    if (p == NULL) {
        return -1;
    }
    inet_ntop(p->ai_family, get_in_addr((struct sockaddr *)p->ai_addr), s, sizeof s);
    
    freeaddrinfo(servinfo); // all done with this structure
    
    // Set socket to ignore SIGPIPE
    int yes = 1;
    setsockopt(sockfd, SOL_SOCKET, SO_NOSIGPIPE, &yes,sizeof(yes));
    
    return sockfd;
}

int swocket_close(int sockfd) {
    return close(sockfd);
}

int swocket_listen(const char * port, int backlog) {
    int sockfd;  // listen on sock_fd, new connection on new_fd
    struct addrinfo hints, *servinfo, *p;
    struct sockaddr_storage; // connector's address information
    struct sigaction sa;
    int yes=1;
    int rv;
    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags = AI_PASSIVE; // use my IP
    if ((rv = getaddrinfo(NULL, port, &hints, &servinfo)) != 0) {
        return -1;
    }
    
    // loop through all the results and bind to the first we can
    for(p = servinfo; p != NULL; p = p->ai_next) {
        if ((sockfd = socket(p->ai_family, p->ai_socktype,
                             p->ai_protocol)) == -1) {
            continue;
        }
        if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &yes,
                       sizeof(int)) == -1) {
            perror("setsockopt");
            return -1;
        }
        if (bind(sockfd, p->ai_addr, p->ai_addrlen) == -1) {
            close(sockfd);
            perror("server: bind");
            continue;
        }
        break;
    }
    if (p == NULL)  {
        return -1;
    }
    freeaddrinfo(servinfo); // all done with this structure
    if (listen(sockfd, backlog) == -1) {
        perror("listen");
        return -1;
    }
    
    sa.sa_handler = sigchld_handler; // reap all dead processes
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = SA_RESTART;
    if (sigaction(SIGCHLD, &sa, NULL) == -1) {
        perror("sigaction");
        return -1;
    }
    
    return sockfd;
}

int swocket_accept(int sockfd) {
    int new_fd;  // listen on sock_fd, new connection on new_fd
    struct sockaddr_storage their_addr; // connector's address information
    socklen_t sin_size;
    char s[INET6_ADDRSTRLEN];
    
    sin_size = sizeof their_addr;
    new_fd = accept(sockfd, (struct sockaddr *)&their_addr, &sin_size);
    
    inet_ntop(their_addr.ss_family,
              get_in_addr((struct sockaddr *)&their_addr),
              s, sizeof s);
    
    return new_fd;
}

int swocket_listen_udp(const char * port) {
    int sockfd;
    struct addrinfo hints, *servinfo, *p;
    int rv;

    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC; // set to AF_INET to force IPv4
    hints.ai_socktype = SOCK_DGRAM;
    hints.ai_flags = AI_PASSIVE; // use my IP
    
    if ((rv = getaddrinfo(NULL, port, &hints, &servinfo)) != 0) {
        return -1;
    }
    // loop through all the results and bind to the first we can
    for(p = servinfo; p != NULL; p = p->ai_next) {
        if ((sockfd = socket(p->ai_family, p->ai_socktype,
                             p->ai_protocol)) == -1) {
            perror("listener: socket");
            continue; }
        if (bind(sockfd, p->ai_addr, p->ai_addrlen) == -1) {
            close(sockfd);
            perror("listener: bind");
            continue; }
        break; }
    if (p == NULL) {
        return -1;
    }
    
    freeaddrinfo(servinfo);
    
    return sockfd;
}

int swocket_connect_udp(const char * host, const char * port) {
    int sockfd;
    struct addrinfo hints, *servinfo, *p;
    int rv;

    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_DGRAM;
    if ((rv = getaddrinfo(host, port, &hints, &servinfo)) != 0) {
        return -1;
    }
    // loop through all the results and make a socket
    for(p = servinfo; p != NULL; p = p->ai_next) {
        if ((sockfd = socket(p->ai_family, p->ai_socktype,
                             p->ai_protocol)) == -1) {
            perror("talker: socket");
            continue;
        }
        break;
    }
    
    if (p == NULL) {
        return -1;
    }
    
    freeaddrinfo(servinfo);
    
    return sockfd;
}

//ssize_t swocket_recieve_udp(int sockfd) {
//    int MAXBUFLEN = 100;
//    
//    ssize_t numbytes;
//    struct sockaddr_storage their_addr;
//    char buf[MAXBUFLEN];
//    socklen_t addr_len;
//    
//    addr_len = sizeof their_addr;
//    if ((numbytes = recvfrom(sockfd, buf, MAXBUFLEN-1 , 0,
//                             (struct sockaddr *)&their_addr, &addr_len)) == -1) {
//        return -1;
//    }
//
//    return numbytes;
//}
//
//ssize_t swocket_send_udp(int sockfd) {
//    int numbytes;
//    
//    if ((numbytes = sendto(sockfd, argv[2], strlen(argv[2]), 0,
//                           p->ai_addr, p->ai_addrlen)) == -1) {
//        return -1;
//    }
//    
//    return 0;
//}
